cpp/src/reductions/reductions.cu

```diff
@@ -47,222 +45,159 @@ void gpu_reduction_op(const T *data, const gdf_valid_type *mask,
+    {
+        using UnderlyingType = typename T::value_type;
+        return launch<UnderlyingType>(col, dev_result);
     }
 
-    template<typename T>
-    static constexpr T identity() { return std::numeric_limits<T>::max(); }
+    template <typename T, typename std::enable_if<
+                  !std::is_arithmetic<T>::value && !is_nonarithmetic_op
+              >::type* = nullptr>
+    gdf_error operator()(gdf_column *col, void *dev_result)
+    {
+        return GDF_UNSUPPORTED_DTYPE;
+    }
 };
 
-struct DeviceMax {
-    typedef IdentityLoader Loader;
-    typedef DeviceMax second;
 
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs >= rhs? lhs: rhs;
-    }
+}   // anonymous namespace
 
-    template<typename T>
-    static constexpr T identity() { return std::numeric_limits<T>::lowest(); }
-};
+typedef enum {
+  GDF_REDUCTION_SUM = 0,
+  GDF_REDUCTION_MIN,
+  GDF_REDUCTION_MAX,
+  GDF_REDUCTION_PRODUCTION,
+  GDF_REDUCTION_SUMOFSQUARES,
+} gdf_reduction_op;
 
-template <typename Op>
-struct ReduceDispatcher {
-    template <typename T,
-              typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
-    gdf_error operator()(gdf_column *col, 
-                         void *dev_result, 
-                         gdf_size_type dev_result_size) {
-        GDF_REQUIRE(col->size > col->null_count, GDF_DATASET_EMPTY);
-        T identity = Op::template identity<T>();
-        return ReduceOp<T, Op>::launch(col, identity, 
-                                       reinterpret_cast<T*>(dev_result), 
-                                       dev_result_size); 
-    }
 
-    template <typename T,
-              typename std::enable_if_t<!std::is_arithmetic<T>::value, T>* = nullptr>
-    gdf_error operator()(gdf_column *col, 
-                         void *dev_result, 
-                         gdf_size_type dev_result_size) {
-        return GDF_UNSUPPORTED_DTYPE;
+gdf_error gdf_reduction(gdf_column *col,
+                  gdf_reduction_op op,
+                  void *dev_result)
+{
+    switch(op){
+    case GDF_REDUCTION_SUM:
+        return cudf::type_dispatcher(col->dtype,
+            ReduceDispatcher<DeviceSum>(), col, dev_result);
+    case GDF_REDUCTION_MIN:
+        return cudf::type_dispatcher(col->dtype,
+            ReduceDispatcher<DeviceMin>(), col, dev_result);
+    case GDF_REDUCTION_MAX:
+        return cudf::type_dispatcher(col->dtype,
+            ReduceDispatcher<DeviceMax>(), col, dev_result);
+    case GDF_REDUCTION_PRODUCTION:
+        return cudf::type_dispatcher(col->dtype,
+            ReduceDispatcher<DeviceProduct>(), col, dev_result);
+    case GDF_REDUCTION_SUMOFSQUARES:
+        return cudf::type_dispatcher(col->dtype,
+            ReduceDispatcher<DeviceSumOfSquares>(), col, dev_result);
+    default:
+        return GDF_INVALID_API_CALL;
     }
-};
-
+}
 
+// ToDo: remove these APIs
```

Comment:
This comment is confusing to whomever is reading the code.. Suggest dropping it in favor of filing an issue.

Reply:
Sorry for confusion.
Will remove the comment.